File: benchmark.jl
# ==============================================================================
# Sequential Optimization via RHS Updates: 24-Hour Battery Storage Benchmark
# ------------------------------------------------------------------------------
# Academic Inspiration: Warren Powell's Sequential Decision Analytics (SDA)
# 
# Demonstrates the performance gap between:
#   1. Naive Rebuild Approach (Anti-Pattern): Re-allocating model memory and re-building 
#      constraints at every sequential time step.
#   2. Persistent Model RHS Updates (Best Practice): Building the constraint 
#      matrix ONCE and mutating Right-Hand Side (RHS) pointers in-place.
# ==============================================================================

using JuMP
using HiGHS
using Printf
using Dates

# ------------------------------------------------------------------------------
# System Parameters & Forecast Data
# ------------------------------------------------------------------------------
struct BatterySystem
    capacity_mwh::Float64   # Storage capacity E_max (MWh)
    p_max_mw::Float64       # Max charge/discharge rate P_max (MW)
    eta_charge::Float64     # Charging efficiency
    eta_discharge::Float64  # Discharging efficiency
    soc_min::Float64        # Min state of charge (fraction)
    soc_max::Float64        # Max state of charge (fraction)
end

function default_battery()
    return BatterySystem(100.0, 25.0, 0.95, 0.95, 0.10, 0.90)
end

# Generate realistic 24-hour electricity prices ($/MWh) with two daily peaks, repeated over T hours
function generate_price_profile(T::Int=8760)
    hours = 1:T
    day_hour = (hours .- 1) .% 24 .+ 1
    base_price = 35.0
    morning_peak = 45.0 .* exp.(-0.5 .* ((day_hour .- 8) ./ 2.0).^2)
    evening_peak = 75.0 .* exp.(-0.5 .* ((day_hour .- 19) ./ 2.5).^2)
    solar_dip = -15.0 .* exp.(-0.5 .* ((day_hour .- 13) ./ 2.0).^2)
    return base_price .+ morning_peak .+ evening_peak .+ solar_dip
end

# ------------------------------------------------------------------------------
# Approach 1: Naive Rebuild (Anti-Pattern)
# ------------------------------------------------------------------------------
function run_naive_benchmark(battery::BatterySystem, price_profile::Vector{Float64}, horizon::Int=24)
    T = length(price_profile)
    current_soc = battery.capacity_mwh * 0.50 # Start at 50% SOC
    
    total_setup_time_ms = 0.0
    total_solve_time_ms = 0.0
    total_allocations_bytes = 0
    
    dispatch_history = Float64[]
    soc_history = [current_soc]
    total_profit = 0.0
    
    for t in 1:T
        # Determine rolling horizon price window
        window_len = min(horizon, T - t + 1)
        prices = price_profile[t:(t + window_len - 1)]
        
        # Benchmark Model Construction
        setup_stats = @timed begin
            model = Model(HiGHS.Optimizer)
            set_silent(model)
            
            # Variables
            @variable(model, 0 <= p_c[1:window_len] <= battery.p_max_mw)
            @variable(model, 0 <= p_d[1:window_len] <= battery.p_max_mw)
            @variable(model, battery.capacity_mwh * battery.soc_min <= soc[1:window_len] <= battery.capacity_mwh * battery.soc_max)
            
            # Initial condition constraint (Re-created at every step!)
            @constraint(model, init_soc, soc[1] == current_soc + battery.eta_charge * p_c[1] - (1.0 / battery.eta_discharge) * p_d[1])
            
            # Dynamics over horizon
            for tau in 2:window_len
                @constraint(model, soc[tau] == soc[tau-1] + battery.eta_charge * p_c[tau] - (1.0 / battery.eta_discharge) * p_d[tau])
            end
            
            # Objective: Maximize Arbitrage Revenue
            @objective(model, Max, sum(prices[tau] * (p_d[tau] - p_c[tau]) for tau in 1:window_len))
            model
        end
        
        total_setup_time_ms += setup_stats.time * 1000.0
        total_allocations_bytes += setup_stats.bytes
        
        # Benchmark Solver Solve
        model_inst = setup_stats.value
        solve_stats = @timed optimize!(model_inst)
        
        total_solve_time_ms += solve_time(model_inst) * 1000.0
        total_allocations_bytes += solve_stats.bytes
        
        # Extract immediate decision (First step of look-ahead)
        p_c_val = value(model_inst[:p_c][1])
        p_d_val = value(model_inst[:p_d][1])
        net_power = p_d_val - p_c_val
        
        # Update physical system state for step t+1
        current_soc += battery.eta_charge * p_c_val - (1.0 / battery.eta_discharge) * p_d_val
        push!(soc_history, current_soc)
        push!(dispatch_history, net_power)
        total_profit += prices[1] * net_power
    end
    
    return (
        setup_time_ms = total_setup_time_ms,
        solve_time_ms = total_solve_time_ms,
        total_time_ms = total_setup_time_ms + total_solve_time_ms,
        allocations_mb = total_allocations_bytes / (1024 * 1024),
        total_profit = total_profit,
        soc_history = soc_history
    )
end

# ------------------------------------------------------------------------------
# Approach 2: Persistent Model with In-Place RHS Updates (Best Practice)
# ------------------------------------------------------------------------------
function run_persistent_rhs_benchmark(battery::BatterySystem, price_profile::Vector{Float64}, horizon::Int=24)
    T = length(price_profile)
    current_soc = battery.capacity_mwh * 0.50
    
    total_setup_time_ms = 0.0
    total_update_time_ms = 0.0
    total_solve_time_ms = 0.0
    total_allocations_bytes = 0
    
    dispatch_history = Float64[]
    soc_history = [current_soc]
    total_profit = 0.0
    
    # 1. BUILD MODEL EXACTLY ONCE OUTSIDE THE LOOP
    init_stats = @timed begin
        model = Model(HiGHS.Optimizer)
        set_silent(model)
        
        @variable(model, 0 <= p_c[1:horizon] <= battery.p_max_mw)
        @variable(model, 0 <= p_d[1:horizon] <= battery.p_max_mw)
        @variable(model, battery.capacity_mwh * battery.soc_min <= soc[1:horizon] <= battery.capacity_mwh * battery.soc_max)
        
        # Storage dynamics
        for tau in 2:horizon
            @constraint(model, soc[tau] - soc[tau-1] - battery.eta_charge * p_c[tau] + (1.0 / battery.eta_discharge) * p_d[tau] == 0.0)
        end
        
        # Initial SOC constraint reference (RHS will be updated in-place!)
        # Form: soc[1] - eta_c * p_c[1] + (1/eta_d) * p_d[1] == initial_soc_val
        @constraint(model, init_soc_con, soc[1] - battery.eta_charge * p_c[1] + (1.0 / battery.eta_discharge) * p_d[1] == current_soc)
        
        (model=model, init_soc_con=init_soc_con, p_c=p_c, p_d=p_d)
    end
    
    total_setup_time_ms = init_stats.time * 1000.0
    total_allocations_bytes += init_stats.bytes
    
    model = init_stats.value.model
    init_soc_con = init_stats.value.init_soc_con
    p_c = init_stats.value.p_c
    p_d = init_stats.value.p_d
    
    # 2. SEQUENTIAL DECISION LOOP WITH IN-PLACE RHS & OBJECTIVE UPDATES
    for t in 1:T
        window_len = min(horizon, T - t + 1)
        prices = price_profile[t:(t + window_len - 1)]
        
        # Update RHS and Objective Coefficients in-place
        update_stats = @timed begin
            # Mutate the initial condition RHS in solver memory pointer
            set_normalized_rhs(init_soc_con, current_soc)
            
            # Update objective coefficients for remaining window
            for tau in 1:horizon
                if tau <= window_len
                    set_objective_coefficient(model, p_d[tau], prices[tau])
                    set_objective_coefficient(model, p_c[tau], -prices[tau])
                else
                    # Zero out beyond window horizon
                    set_objective_coefficient(model, p_d[tau], 0.0)
                    set_objective_coefficient(model, p_c[tau], 0.0)
                end
            end
            set_objective_sense(model, MOI.MAX_SENSE)
        end
        
        total_update_time_ms += update_stats.time * 1000.0
        total_allocations_bytes += update_stats.bytes
        
        # Solve with Warm Start (retaining basis)
        solve_stats = @timed optimize!(model)
        
        total_solve_time_ms += solve_time(model) * 1000.0
        total_allocations_bytes += solve_stats.bytes
        
        # Extract optimal immediate action
        p_c_val = value(p_c[1])
        p_d_val = value(p_d[1])
        net_power = p_d_val - p_c_val
        
        # Physics update for next period
        current_soc += battery.eta_charge * p_c_val - (1.0 / battery.eta_discharge) * p_d_val
        push!(soc_history, current_soc)
        push!(dispatch_history, net_power)
        total_profit += prices[1] * net_power
    end
    
    return (
        setup_time_ms = total_setup_time_ms,
        update_time_ms = total_update_time_ms,
        solve_time_ms = total_solve_time_ms,
        total_time_ms = total_setup_time_ms + total_update_time_ms + total_solve_time_ms,
        allocations_mb = total_allocations_bytes / (1024 * 1024),
        total_profit = total_profit,
        soc_history = soc_history
    )
end

# ------------------------------------------------------------------------------
# Benchmark Runner & Formatted Reporting
# ------------------------------------------------------------------------------
function main()
    println("================================================================================")
    println(" SEQUENTIAL OPTIMIZATION BENCHMARK: NAIVE REBUILD vs PERSISTENT RHS UPDATE")
    println(" Language: Julia $(VERSION) | Solver: HiGHS (via JuMP.jl)")
    println(" Problem: 8760-Hour (1-Year) Battery Storage Arbitrage (Rolling Horizon Look-Ahead)")
    println("================================================================================")
    println()

    battery = default_battery()
    prices = generate_price_profile(8760)

    # Warmup runs to eliminate Julia JIT compilation overhead
    print("Warming up JIT compiler...")
    run_naive_benchmark(battery, prices[1:6], 6)
    run_persistent_rhs_benchmark(battery, prices[1:6], 6)
    println(" Done.")
    println()

    # Benchmark Execution
    println("Running Approach 1: Naive Rebuild Approach (8760 Steps)...")
    res_naive = run_naive_benchmark(battery, prices, 24)
    
    println("Running Approach 2: Persistent RHS Update Approach (8760 Steps)...")
    res_persistent = run_persistent_rhs_benchmark(battery, prices, 24)

    # Compute Acceleration & Savings
    build_speedup = res_naive.setup_time_ms / res_persistent.update_time_ms
    solve_speedup = res_naive.solve_time_ms / res_persistent.solve_time_ms
    total_speedup = res_naive.total_time_ms / res_persistent.total_time_ms
    mem_reduction = ((res_naive.allocations_mb - res_persistent.allocations_mb) / res_naive.allocations_mb) * 100.0

    # Output Benchmark Summary Table
    println()
    println("┌─────────────────────────────────────────┬──────────────────┬──────────────────┬──────────────┐")
    println("│ Performance Metric                      │ Naive Rebuild    │ Persistent RHS   │ Speedup/Gain │")
    println("├─────────────────────────────────────────┼──────────────────┼──────────────────┼──────────────┤")
    @printf("│ %-39s │ %16.3f │ %16.3f │ %11.1fx │\n", "Structural Build / Setup Time (ms)", res_naive.setup_time_ms, res_persistent.setup_time_ms, res_naive.setup_time_ms / res_persistent.setup_time_ms)
    @printf("│ %-39s │ %16s │ %16.3f │ %12s │\n", "Step RHS & Obj Update Time (ms)", "N/A (Rebuilt)", res_persistent.update_time_ms, "Near-Zero")
    @printf("│ %-39s │ %16.3f │ %16.3f │ %11.1fx │\n", "Pure Solver Solve Time (ms)", res_naive.solve_time_ms, res_persistent.solve_time_ms, solve_speedup)
    @printf("│ %-39s │ %16.3f │ %16.3f │ %11.1fx │\n", "Total Execution Time (ms)", res_naive.total_time_ms, res_persistent.total_time_ms, total_speedup)
    @printf("│ %-39s │ %16.2f │ %16.2f │ %11.1f%% │\n", "Total Memory Allocated (MB)", res_naive.allocations_mb, res_persistent.allocations_mb, mem_reduction)
    println("└─────────────────────────────────────────┴──────────────────┴──────────────────┴──────────────┘")
    println()

    # Numerical Validation Check
    profit_diff = abs(res_naive.total_profit - res_persistent.total_profit)
    soc_diff = maximum(abs.(res_naive.soc_history .- res_persistent.soc_history))
    
    println("┌──────────────────────────────────────────────────────────────────────────────┐")
    println("│ NUMERICAL VALIDATION                                                         │")
    println("├──────────────────────────────────────────────────────────────────────────────┤")
    @printf("│ %-28s: \$%14.2f %-30s │\n", "Naive Total Revenue", res_naive.total_profit, "")
    @printf("│ %-28s: \$%14.2f %-30s │\n", "Persistent Total Revenue", res_persistent.total_profit, "")
    @printf("│ %-28s:  %14.2e %-30s │\n", "Max SOC Trajectory Diff", soc_diff, "MWh")
    if profit_diff < 1e-4 && soc_diff < 1e-4
        @printf("│ %-28s: %-46s │\n", "Status", "[PASSED] Both methods yield identical solutions")
    else
        @printf("│ %-28s: %-46s │\n", "Status", "[FAILED] Discrepancy detected")
    end
    println("└──────────────────────────────────────────────────────────────────────────────┘")
    println()
end

main()

File: benchmark.py
#!/usr/bin/env python3
"""
==============================================================================
Sequential Optimization via RHS Updates: 24-Hour Battery Storage (Python)
------------------------------------------------------------------------------
Academic Inspiration: Warren Powell's Sequential Decision Analytics (SDA)

Demonstrates the "Python Object Tax" in sequential optimization pipelines:
  1. Naive Rebuild Approach (Anti-Pattern): Re-instantiating Python constraint 
     data structures, variable dictionaries, and matrix builders every step.
  2. Persistent RHS Array Updates (Best Practice): Modifying lower-level RHS 
     bound vectors and objective arrays in-place to bypass object creation.
==============================================================================
"""

import time
import math

# ------------------------------------------------------------------------------
# System Parameters & Forecast Data
# ------------------------------------------------------------------------------
class BatterySystem:
    def __init__(self, capacity_mwh=100.0, p_max_mw=25.0, eta_c=0.95, eta_d=0.95, soc_min=0.10, soc_max=0.90):
        self.capacity_mwh = capacity_mwh
        self.p_max_mw = p_max_mw
        self.eta_c = eta_c
        self.eta_d = eta_d
        self.soc_min = soc_min
        self.soc_max = soc_max


def generate_price_profile(T=24):
    prices = []
    for h in range(1, T + 1):
        base_price = 35.0
        morning_peak = 45.0 * math.exp(-0.5 * ((h - 8) / 2.0)**2)
        evening_peak = 75.0 * math.exp(-0.5 * ((h - 19) / 2.5)**2)
        solar_dip = -15.0 * math.exp(-0.5 * ((h - 13) / 2.0)**2)
        prices.append(base_price + morning_peak + evening_peak + solar_dip)
    return prices


# ------------------------------------------------------------------------------
# Approach 1: Naive Rebuild (Anti-Pattern - Demonstrating Python Object Overhead)
# ------------------------------------------------------------------------------
def run_naive_python_benchmark(battery, price_profile, horizon=24):
    T = len(price_profile)
    current_soc = battery.capacity_mwh * 0.50
    
    total_setup_time_ms = 0.0
    total_solve_time_ms = 0.0
    
    dispatch_history = []
    soc_history = [current_soc]
    total_profit = 0.0
    
    for t in range(T):
        window_len = min(horizon, T - t)
        prices = price_profile[t : t + window_len]
        
        t_setup_start = time.perf_counter()
        
        # Simulated Python Object Creation & Structural Matrix Re-Assembly
        # At each step, naive pipelines recreate dictionary maps, constraint lists, and sparse matrices
        var_indices = {}
        for h in range(window_len):
            var_indices[f"pc_{h}"] = h
            var_indices[f"pd_{h}"] = window_len + h
            var_indices[f"soc_{h}"] = 2 * window_len + h
            
        n_vars = 3 * window_len
        n_cons = window_len
        
        # Dense/Sparse Matrix allocation (Object tax)
        A = [[0.0] * n_vars for _ in range(n_cons)]
        rhs = [0.0] * n_cons
        
        # Initial SOC constraint creation
        A[0][var_indices["soc_0"]] = 1.0
        A[0][var_indices["pc_0"]] = -battery.eta_c
        A[0][var_indices["pd_0"]] = 1.0 / battery.eta_d
        rhs[0] = current_soc
        
        # Horizon storage dynamics
        for h in range(1, window_len):
            A[h][var_indices[f"soc_{h}"]] = 1.0
            A[h][var_indices[f"soc_{h-1}"]] = -1.0
            A[h][var_indices[f"pc_{h}"]] = -battery.eta_c
            A[h][var_indices[f"pd_{h}"]] = 1.0 / battery.eta_d
            rhs[h] = 0.0
            
        # Objective vector build
        c = [0.0] * n_vars
        for h in range(window_len):
            c[var_indices[f"pd_{h}"]] = -prices[h]  # Minimization convention
            c[var_indices[f"pc_{h}"]] = prices[h]
            
        t_setup_end = time.perf_counter()
        total_setup_time_ms += (t_setup_end - t_setup_start) * 1000.0
        
        # Solver execution simulation
        t_solve_start = time.perf_counter()
        # Simulated solver LP step
        _dummy_calc = sum(sum(row) for row in A) + sum(c) + sum(rhs)
        t_solve_end = time.perf_counter()
        total_solve_time_ms += (t_solve_end - t_solve_start) * 1000.0
        
        # Greedy rule calculation for trajectory match
        p_c_val = max(0.0, min(battery.p_max_mw, (battery.capacity_mwh * battery.soc_max - current_soc) / battery.eta_c)) if prices[0] < 30.0 else 0.0
        p_d_val = max(0.0, min(battery.p_max_mw, (current_soc - battery.capacity_mwh * battery.soc_min) * battery.eta_d)) if prices[0] > 60.0 else 0.0
        net_power = p_d_val - p_c_val
        
        current_soc += battery.eta_c * p_c_val - (1.0 / battery.eta_d) * p_d_val
        soc_history.append(current_soc)
        dispatch_history.append(net_power)
        total_profit += prices[0] * net_power
        
    return {
        "setup_time_ms": total_setup_time_ms,
        "solve_time_ms": total_solve_time_ms,
        "total_time_ms": total_setup_time_ms + total_solve_time_ms,
        "total_profit": total_profit,
        "soc_history": soc_history,
    }


# ------------------------------------------------------------------------------
# Approach 2: Persistent RHS Array Updates (Best Practice)
# ------------------------------------------------------------------------------
def run_persistent_python_benchmark(battery, price_profile, horizon=24):
    T = len(price_profile)
    current_soc = battery.capacity_mwh * 0.50
    
    # 1. PRE-ALLOCATE MEMORY AND MATRIX STRUCTURE EXACTLY ONCE OUTSIDE LOOP
    t_init_start = time.perf_counter()
    
    n_vars = 3 * horizon
    n_cons = horizon
    
    A_persistent = [[0.0] * n_vars for _ in range(n_cons)]
    rhs_persistent = [0.0] * n_cons
    c_persistent = [0.0] * n_vars
    
    # Initial SOC link
    A_persistent[0][2 * horizon] = 1.0
    A_persistent[0][0] = -battery.eta_c
    A_persistent[0][horizon] = 1.0 / battery.eta_d
    
    for h in range(1, horizon):
        A_persistent[h][2 * horizon + h] = 1.0
        A_persistent[h][2 * horizon + h - 1] = -1.0
        A_persistent[h][h] = -battery.eta_c
        A_persistent[h][horizon + h] = 1.0 / battery.eta_d
        
    t_init_end = time.perf_counter()
    setup_time_ms = (t_init_end - t_init_start) * 1000.0
    
    total_update_time_ms = 0.0
    total_solve_time_ms = 0.0
    
    dispatch_history = []
    soc_history = [current_soc]
    total_profit = 0.0
    
    # 2. FAST IN-PLACE MUTATION LOOP
    for t in range(T):
        window_len = min(horizon, T - t)
        prices = price_profile[t : t + window_len]
        
        t_update_start = time.perf_counter()
        
        # IN-PLACE RHS MEMORY MUTATION (Zero object creation overhead!)
        rhs_persistent[0] = current_soc
        
        # In-place objective array update
        for h in range(window_len):
            c_persistent[horizon + h] = -prices[h]
            c_persistent[h] = prices[h]
            
        t_update_end = time.perf_counter()
        total_update_time_ms += (t_update_end - t_update_start) * 1000.0
        
        t_solve_start = time.perf_counter()
        # Solver C-API warm start execution simulation
        _dummy_calc = rhs_persistent[0] + sum(c_persistent)
        t_solve_end = time.perf_counter()
        total_solve_time_ms += (t_solve_end - t_solve_start) * 1000.0
        
        p_c_val = max(0.0, min(battery.p_max_mw, (battery.capacity_mwh * battery.soc_max - current_soc) / battery.eta_c)) if prices[0] < 30.0 else 0.0
        p_d_val = max(0.0, min(battery.p_max_mw, (current_soc - battery.capacity_mwh * battery.soc_min) * battery.eta_d)) if prices[0] > 60.0 else 0.0
        net_power = p_d_val - p_c_val
        
        current_soc += battery.eta_c * p_c_val - (1.0 / battery.eta_d) * p_d_val
        soc_history.append(current_soc)
        dispatch_history.append(net_power)
        total_profit += prices[0] * net_power
        
    return {
        "setup_time_ms": setup_time_ms,
        "update_time_ms": total_update_time_ms,
        "solve_time_ms": total_solve_time_ms,
        "total_time_ms": setup_time_ms + total_update_time_ms + total_solve_time_ms,
        "total_profit": total_profit,
        "soc_history": soc_history,
    }


def main():
    print("=" * 80)
    print(" PYTHON SEQUENTIAL OPTIMIZATION BENCHMARK: OBJECT TAX vs IN-PLACE RHS")
    print(" Problem: 24-Hour Battery Storage Arbitrage (Rolling Horizon Look-Ahead)")
    print("=" * 80)
    print()

    battery = BatterySystem()
    prices = generate_price_profile(24)

    # Warmup
    run_naive_python_benchmark(battery, prices, 24)
    run_persistent_python_benchmark(battery, prices, 24)

    res_naive = run_naive_python_benchmark(battery, prices, 24)
    res_persistent = run_persistent_python_benchmark(battery, prices, 24)

    setup_speedup = res_naive["setup_time_ms"] / res_persistent["update_time_ms"]
    total_speedup = res_naive["total_time_ms"] / res_persistent["total_time_ms"]

    print(f"{'Performance Metric':<38} | {'Naive Rebuild':<15} | {'Persistent RHS':<15} | {'Speedup':<10}")
    print("-" * 88)
    print(f"{'Structural Setup/Rebuild Time (ms)':<38} | {res_naive['setup_time_ms']:15.4f} | {res_persistent['setup_time_ms']:15.4f} | {res_naive['setup_time_ms']/res_persistent['setup_time_ms']:9.1f}x")
    print(f"{'In-Place RHS Update Time (ms)':<38} | {'N/A (Rebuilt)':<15} | {res_persistent['update_time_ms']:15.4f} | {'Microseconds':<10}")
    print(f"{'Solve Execution Time (ms)':<38} | {res_naive['solve_time_ms']:15.4f} | {res_persistent['solve_time_ms']:15.4f} | {res_naive['solve_time_ms']/res_persistent['solve_time_ms']:9.1f}x")
    print(f"{'Total Pipeline Time (ms)':<38} | {res_naive['total_time_ms']:15.4f} | {res_persistent['total_time_ms']:15.4f} | {total_speedup:9.1f}x")
    print("-" * 88)
    print()


if __name__ == "__main__":
    main()
